### PR TITLE
Feature/refactor program

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
This change does away with use of deprecated ``sycl::program`` class replacing it with use of ``sycl::kernel_bundle<sycl::bundle_state::executable>`` instead.

`DPCTLSyclProgramRef` has been removed, but `DPCTLSyclKernelBundleRef` introduced, being an opaque reference to the executable kernel bundle as stated above.

Functions `DPCTLProgram_*` are renamed to `DPCTLKernelBundle_*`. 

Free functions `DPCTLKernelBundle_Create*` now take context and device references (previously only took `DPCTLSyclContextRef` argument).

Implementation of `dpctl_sycl_program_interface` no longer directly uses OpenCL symbols, uses dynamic_Lib helper class to load them dynamically. Thus SyclInterface library no longer links with OpenCL library.

- [X] Have you provided a meaningful PR description?
- [X] Have you added a test, reproducer or referred to an issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

**Comment by oleksandr-pavlyk** [comment]
@diptorupd As far as I can see this change show not require companion changes in numba-dpex.

File: libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp [hunk 1]
```diff
@@ -0,0 +1,784 @@
...
+const char *zeKernelCreate_Name = "zeKernelCreate";
+zeKernelCreateFT get_zeKernelCreate()
+{
+    static auto st_zeKernelCreateF =
+        ze_loader::get().getSymbol<zeKernelCreateFT>(zeKernelCreate_Name);
+
+    return st_zeKernelCreateF;
+}
+
+std::string _GetErrorCode_ze_impl(ze_result_t code)
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp:427]: Same here. Not all `ze_result_t` codes are checked. Only those that used Level-Zero functions are documented to return. Perhaps using `switch` is better than chaining `if`/`else`.

File: libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp [hunk 2]
```diff
@@ -0,0 +1,784 @@
...
+                      __FILE__, __func__, __LINE__);
+        return nullptr;
+    }
+    static auto st_clCreateKernelF =
+        clLib.getSymbol<clCreateKernelFT>(clCreateKernel_Name);
+
+    return st_clCreateKernelF;
+}
+
+std::string _GetErrorCode_ocl_impl(cl_int code)
```
[the remark this answers is not in the document]
Comment (oleksandr-pavlyk) [reply libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp:163]: I have pushed a change to use `switch` over chained `if`/`else` clauses.

**Comment by oleksandr-pavlyk** [comment]
Functions `DPCTLKernelBundle_Create*` dispatch to implementations for respective backends. Backend entity is created (either `cl_program` or `ze_module_handle_t`), which are used in `sycl::make_kernel_bundle` to create the interoperability kernel bundle. 

Since interoperability kernel_bundle does not expose kernel ids for kernels in it, `GetKernel` and `HasKernel` functions are implementing by getting the native component out. This native component is a vector of `cl_program` or `ze_module_handle_t`. 

For each of this the backend-appropriate function is used to look-up the kernel by its name. 

SYCL does not guarantee uniqueness of the kernel with the given name in a `kernel_bundle`. The functionality retrieves the first match found. 

Kernel bundles constructed from OCL source or from a single SPV are guaranteed to not have duplicates.

File: dpctl/program/_program.pxd [hunk 3]
```diff
@@ -33,26 +33,26 @@ cdef class SyclKernel:
...
 
     @staticmethod
-    cdef SyclKernel _create (DPCTLSyclKernelRef kref)
+    cdef SyclKernel _create (DPCTLSyclKernelRef kref, str name)
 
 
 cdef class SyclProgram:
-    ''' Wraps a sycl::program object created from an OpenCL interoperability
-        program.
+    ''' Wraps a sycl::kernel_bundle<sycl::bundle_state::executable> object created from
```
Comment (diptorupd) [inline dpctl/program/_program.pxd:44]: Something is off with grammar "created from using"

File: dpctl/program/_program.pxd [hunk 4]
```diff
@@ -33,26 +33,26 @@ cdef class SyclKernel:
         kernel.
     '''
     cdef DPCTLSyclKernelRef _kernel_ref
-    cdef const char *_function_name
+    cdef str _function_name
```
Comment (diptorupd) [inline dpctl/program/_program.pxd:36]: Why the change? It is for my education about Cython mostly.

Comment (oleksandr-pavlyk) [reply dpctl/program/_program.pxd:36]: Mostly to delegate copying and memory management to Python, which simplifies the code. No other reason.

File: dpctl/program/_program.pxd (hunk 3 above)
Comment (oleksandr-pavlyk) [reply dpctl/program/_program.pxd:44]: Yes, should be "created by using".

Comment (oleksandr-pavlyk) [reply dpctl/program/_program.pxd]: Pushed the change. Resolving.

File: dpctl/tests/test_sycl_program.py [hunk 5]
```diff
@@ -31,6 +31,24 @@ def get_spirv_abspath(fn):
     return spirv_file
 
 
+def _check_multi_kernel_program(prog):
+    assert type(prog) is dpctl_prog.SyclProgram
```
Comment (diptorupd) [inline dpctl/tests/test_sycl_program.py:35]: Should we be using `assert`? (Recently concluded Security training speaking here)

File: libsyclinterface/include/Config/dpctl_config.h.in [hunk 6]
```diff
@@ -26,9 +26,11 @@
 #pragma once
 
 /* Defined when dpctl was built with level zero program creation enabled. */
-#cmakedefine DPCTL_ENABLE_L0_PROGRAM_CREATION @DPCTL_ENABLE_L0_PROGRAM_CREATION@
+#cmakedefine DPCTL_ENABLE_L0_PROGRAM_CREATION                                  \
```
Comment (diptorupd) [inline libsyclinterface/include/Config/dpctl_config.h.in:29]: :+1:

File: libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp [hunk 7]
```diff
@@ -0,0 +1,734 @@
...
+
+typedef cl_program (*clCreateProgramWithSourceFT)(cl_context,
+                                                  cl_uint,
+                                                  const char **,
+                                                  const size_t *,
+                                                  cl_int *);
+const char *clCreateProgramWithSource_Name = "clCreateProgramWithSource";
+clCreateProgramWithSourceFT get_clCreateProgramWithSource()
+{
+    static auto st_clCreateProgramWithSourceF =
```
Comment (diptorupd) [inline libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp:111]: Maybe we should add a check here that the symbol was found?

File: libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp [hunk 8]
```diff
@@ -0,0 +1,734 @@
...
+DPCTLSyclKernelBundleRef
+_CreateKernelBundle_common_ocl_impl(cl_program clProgram,
+                                    const context &ctx,
+                                    const device &dev,
+                                    const char *CompileOpts)
+{
+    backend_traits<cl_be>::return_type<device> clDevice;
+    clDevice = get_native<cl_be>(dev);
+
+    // Last to pointers are notification function pointer and user-data pointer
```
Comment (diptorupd) [inline libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp:181]: s/to/two/g

**Review by diptorupd — APPROVED** [review]
@oleksandr-pavlyk Overall it looks good. There may be a way to make the `dpctl_sycl_kernel_bundle_interface.cpp` more modular. E.g. having an interface and then each implementation of the interface `OCL`, `ZE` implementing the interface. But, we may soon find ourselves reinventing the DPC++ PI. So thanks, but no thanks!

Other things, I mostly glanced through so might have missed few things. If you are satisfied please feel free to merge.

File: libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp (hunk 7 above)
Comment (oleksandr-pavlyk) [reply libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp:111]: There is a check already in the `DynamicLibHelper`` method. Additionally there is a check after each call to `get_SymbNameF` function.

File: libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp (hunk 8 above)
Comment (oleksandr-pavlyk) [inline libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp:181]:
```suggestion
    // Last two pointers are notification function pointer and user-data pointer
```

Comment (oleksandr-pavlyk) [reply libsyclinterface/source/dpctl_sycl_kernel_bundle_interface.cpp]: Fixed.
